### PR TITLE
dialects: FSM advanced operator (SW/HW instances)

### PR DESCRIPTION
**Pull request by luisacicolini**
this PR is built on top of #1600 and integrates the base operators and testing with advanced ones useful to emulate FSM behavior in hw or sw.

### CODE REVIEW

**Comment by compor** [comment]
Can you please stack it on top of #1600? it'd be easier to review

File: xdsl/dialects/fsm.py [hunk 1]
```diff
@@ -45,6 +47,11 @@
 signlessIntegerLike = ContainerOf(AnyOf([IntegerType, IndexType]))
 
 
+@irdl_attr_definition
+class InstanceType(ParametrizedAttribute, TypeAttribute):
+    name = "fsm.instancetype"
```
Comment (compor) [inline xdsl/dialects/fsm.py:52]: Could you add some doc of this type, please?

File: xdsl/dialects/fsm.py [hunk 2]
```diff
@@ -30,9 +30,11 @@
     opt_attr_def,
     opt_operand_def,
     region_def,
+    result_def,
     var_operand_def,
     var_result_def,
 )
+from xdsl.irdl.irdl import irdl_attr_definition
```
Comment (compor) [inline xdsl/dialects/fsm.py:37]:
```suggestion
from xdsl.irdl import irdl_attr_definition
```

File: xdsl/dialects/fsm.py [hunk 3]
```diff
@@ -363,6 +372,156 @@ def __init__(
         )
 
 
+@irdl_op_definition
+class Instance(IRDLOperation):
+    """Represents an instance of a state machine, including an
+    instance name and a symbol reference of the machine"""
```
Comment (compor) [inline xdsl/dialects/fsm.py:378]:
```suggestion
    """
    Represents an instance of a state machine, including an
    instance name and a symbol reference of the machine.
    """
    
    Similar in other docstrings.
```

File: xdsl/dialects/fsm.py [hunk 4]
```diff
@@ -363,6 +372,156 @@ def __init__(
         )
 
 
+@irdl_op_definition
+class Instance(IRDLOperation):
```
Comment (compor) [inline xdsl/dialects/fsm.py:376]:
```suggestion
class InstanceOp(IRDLOperation):
```

likewise below.

File: xdsl/dialects/fsm.py [hunk 5]
```diff
@@ -200,11 +209,11 @@ def verify_(self):
             if isinstance(parent, Machine):
                 if not (
                     [operand.type for operand in self.operands]
-                    == [type(result) for result in parent.function_type.outputs]
+                    == [result for result in parent.function_type.outputs]
                     and len(self.operands) == len(parent.function_type.outputs)
                 ):
                     raise VerifyException(
-                        "Output types must be consistent with the machine's"
+                        "Output type must be consistent with the machine's"
```
Comment (compor) [inline xdsl/dialects/fsm.py:216]:
```suggestion
                        "Output type must match with the containing state machine's output types"
```

I know that the doc says "consistent", but that means to be the same/matching, right?
If so, reword messages below too?

File: xdsl/dialects/fsm.py [hunk 6]
```diff
@@ -363,6 +372,156 @@ def __init__(
...
+    def verify_(self):
+        m = SymbolTable.lookup_symbol(self, self.machine)
+        if isinstance(m, Machine):
+            if not (
+                [operand.type for operand in self.inputs]
+                == [result for result in m.function_type.inputs]
+                and len(self.inputs) == len(m.function_type.inputs)
+            ):
+                raise VerifyException(
+                    "Inputs types must be consistent with the machine's "
```
Comment (compor) [inline xdsl/dialects/fsm.py:511]: Could you add the state machine's `sym_name` in these messages? Could it help debugging by knowing to which machine this error refers too based on the symbol table?

If so, adjust in the other similar cases where a symbol table lookup is involved too?

**Review by compor — CHANGES_REQUESTED** [review]

File: xdsl/dialects/fsm.py (hunk 1 above)
Comment (luisacicolini) [reply xdsl/dialects/fsm.py:52]: Unfortunately I was not able to find any. I managed to define it by basically looking at .cpp and .td files, as the documentation refers to it was "FSM instance type" :(

File: xdsl/dialects/fsm.py (hunk 5 above)
Comment (luisacicolini) [reply xdsl/dialects/fsm.py:216]: correct, it basically needs to be the same type as the output type of the function defined in the MachineOp

**Comment by JosseVanDelm** [comment]
Some files look like they were already reviewed in #1600 ?

**Comment by luisacicolini** [comment]
Thank you everyone for your suggestions :) 
I added new more significant strings for exceptions and rebased the branch to main (after merging #1600). 
Please let me know if there are any major changes I should implement!

**Comment by superlopuh** [comment]
There are still some merge conflicts left over, could you please resolve them first?

**Comment by compor** [comment]
@luisacicolini can you ping us when you resolve the conflicts and everything on this patch is properly based on `main` please?

**Comment by luisacicolini** [comment]
Manually checked the files again this morning, should be okay!

File: xdsl/dialects/fsm.py [hunk 7]
```diff
@@ -228,6 +232,8 @@ class TransitionOp(IRDLOperation):
 
     action = region_def()
 
+    # attributes
```
Comment (compor) [inline xdsl/dialects/fsm.py:235]:
```suggestion
```

File: xdsl/dialects/fsm.py [hunk 8]
```diff
@@ -371,6 +372,176 @@ def __init__(
...
+            operands=[inputs, instance],
+            result_types=[outputs],
+        )
+
+    def verify_(self):
+        if not isinstance(self.instance.owner, InstanceOp):
+            raise VerifyException("The instance operand must be Instance")
+
+        # operand types must match the machine input types
+        # result types must match the machine output types
```
Comment (compor) [inline xdsl/dialects/fsm.py:445]:
```suggestion
```

**Review by compor — APPROVED** [review]
LGTM with some nits

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/fsm.py [hunk 9]
```diff
@@ -297,20 +299,22 @@ def verify_(self) -> None:
...
+                        found += 1
+                if found == 0:
+                    raise VerifyException(
+                        "Update must only be located in the action region of a transition"
+                    )
+                elif found > 1:
+                    raise VerifyException(
+                        "Multiple updates to the same variable within a single action region is disallowed"
+                    )
+            parent = parent.parent_op()
```
Comment (math-fehr) [inline xdsl/dialects/fsm.py:317]: I feel that this change is backwards, maybe it's a merge problem?
If you want I could take a look with you tomorrow

File: xdsl/dialects/fsm.py [hunk 10]
```diff
@@ -195,9 +196,11 @@ def __init__(
 
     def verify_(self):
         parent = self.parent_op()
-        assert isinstance(parent, StateOp)
```
Comment (math-fehr) [inline xdsl/dialects/fsm.py:198]: Same here, this feels backwards?

**Review by math-fehr — APPROVED** [review]
Really nice!
I think some of these changes appeared while you were doing your rebase?
Because some of the changes in the earlier operations are actually old code I believe?

File: xdsl/dialects/fsm.py (hunk 9 above)
Comment (luisacicolini) [reply xdsl/dialects/fsm.py:317]: yup I guess you're right, I had quite some issues with main merging-rebasing so I guess I missed that! Sure thing :)

**Pull request merged**